### PR TITLE
#7434: automatically activate deployed mods in Extension Console

### LINKED ISSUE #7434: Automatically activate mod deployments when opening Extension Console
Context
---
- The current behavior is to show a deployment banner (or modal, if deployments are required)
- This is confusing to users who expect deployments to be automatically updated
  - Because the Extension Console doesn't listen for active extension state, the mods may be activated by the background heart beat, but wouldn't be reflected in the UI

Definition of Done
---
- Must Have: on Extension Console load, automatically activate deployments if possible. If not, show the existing UI elements (banner/modal)

Discussion
---
- What would be involved in having the Extension Console listen for changes in the active extensions slice to update the status on the mods screen? There are some UI corner cases, e.g., if you've toggled the 3-dot menu for a mod it will disappear on state update due to the re-render. An option might be to show a UI indicator that active mods have changed, with a reload page button

Implementation Sketch
---
- Modify `useDeployments` hook to first try activating deployments on load: https://github.com/pixiebrix/pixiebrix-extension/blob/76ee8fa649ed741ab941ccab39a796ea0479d273/src/extensionConsole/pages/deployments/DeploymentsContext.tsx#L178-L178

### PR DESCRIPTION
**Pull request by grahamlangford**
## What does this PR do?

- Closes #7434

## Reviewer Tips

- _What order should the reviewer review the files in?_
- _Are there any implementation approaches you want feedback on?_

## Discussion

- _Were there multiple approaches you were deciding between?_

## Demo

- _Paste a screenshot or demo video here_

## Future Work

- _Work for the issue/ticket that will be in a follow-up PR_

## Team Coordination

_Leave all that are relevant and check off as completed_

- [ ] This PR requires security review
- [ ] This PR introduces a new library: double check it's MIT/Apache2/permissively licensed
- [ ] This PR requires a node/npm version update: let the team know on #engineering
- [ ] This PR requires a documentation change (link to old docs)
- [ ] This PR requires a tutorial update (link to old tutorials)
- [ ] This PR requires a feature flag
- [ ] This PR requires a environment variable change

## Checklist

- [ ] Add tests
- [ ] New files added to `src/tsconfig.strictNullChecks.json` (if possible)
- [ ] Designate a primary reviewer

### CODE REVIEW

File: src/extensionConsole/pages/deployments/useAutoDeploy.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import { AnyAction, type Dispatch } from "@reduxjs/toolkit";
+import { useState } from "react";
+import { useDispatch } from "react-redux";
+import useAsyncEffect from "use-async-effect";
+
+function useAutoDeploy(
+  deployments: Deployment[],
+  installedExtensions: ModComponentBase[],
+  isLoadingDeployments: boolean,
+): boolean {
```
Comment (twschiller) [inline src/extensionConsole/pages/deployments/useAutoDeploy.ts:33]: The boolean return value needs documentation on what it means

File: src/extensionConsole/pages/deployments/useAutoDeploy.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  const dispatch = useDispatch<Dispatch<AnyAction>>();
+  const [isAutoDeploying, setIsAutoDeploying] = useState(true);
```
Comment (twschiller) [inline src/extensionConsole/pages/deployments/useAutoDeploy.ts:35]: These variables have similar names, not very obviously what the relationship between them are. Looks like isAttemptingAutoDeploy is being used to avoid running the effect 2x concurrently?

File: src/extensionConsole/pages/deployments/useAutoDeploy.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import notify from "@/utils/notify";
+import { AnyAction, type Dispatch } from "@reduxjs/toolkit";
+import { useState } from "react";
+import { useDispatch } from "react-redux";
+import useAsyncEffect from "use-async-effect";
+
+function useAutoDeploy(
+  deployments: Deployment[],
+  installedExtensions: ModComponentBase[],
+  isLoadingDeployments: boolean,
```
Comment (twschiller) [inline src/extensionConsole/pages/deployments/useAutoDeploy.ts:32]: Prefer options so usage is clearer at callsite. This parameter is basically encoding whether or not deployments should be ignored? Might be easier to have callsite pass in undefined for deployments while deployments are loading?

File: src/extensionConsole/pages/deployments/useAutoDeploy.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+): boolean {
+  const dispatch = useDispatch<Dispatch<AnyAction>>();
+  const [isAutoDeploying, setIsAutoDeploying] = useState(true);
+  const [isAttemptingAutoDeploy, setIsAttemptingAutoDeploy] = useState(false);
+  const { hasPermissions } = useModPermissions(installedExtensions);
+
+  useAsyncEffect(
+    async (isMounted) => {
+      // Still loading deployments or already deploying
+      if (!isMounted || isLoadingDeployments || isAttemptingAutoDeploy) {
```
Comment (twschiller) [inline src/extensionConsole/pages/deployments/useAutoDeploy.ts:42]: IIRC, isMounted is a method. So will never be nullish by itself. (Would be good to introduce a lint rule)

File: src/permissions/permissionsUtils.ts [hunk 5]
```diff
@@ -74,7 +74,7 @@ export function mergePermissionsStatuses(
 
 /**
  * Request any permissions the user has not already granted. Must be called from a user gesture.
- * @returns {Promise<boolean>} true iff the all the permissions already existed, or if the user accepted
+ * @returns {Promise<boolean>} true if the all the permissions already existed, or if the user accepted
```
Comment (twschiller) [inline src/permissions/permissionsUtils.ts:77]: The iff was intentional: https://en.wikipedia.org/wiki/If_and_only_if#:~:text=In%20logic%20and%20related%20fields,true%20or%20both%20are%20false.

File: src/extensionConsole/pages/mods/ModsPageLayout.tsx [hunk 6]
```diff
@@ -162,7 +164,7 @@ const ModsPageLayout: React.FunctionComponent<{
         <ModsPageToolbar tableInstance={tableInstance} />
         {/* This wrapper prevents AutoSizer overflow in a flex box container */}
         <div style={{ flex: "1 1 auto" }}>
-          {isLoading ? (
+          {isLoading || isAutoDeploying ? (
```
Comment (twschiller) [inline src/extensionConsole/pages/mods/ModsPageLayout.tsx:167]: Would be good to leave comment on why it's not OK to show mod page content while auto-deployment is occuring

**Comment by twschiller** [comment]
This approach is looking good. Will need to clean up circular dependency. I left some comments on where naming was hard to follow

File: src/extensionConsole/pages/deployments/useAutoDeploy.ts (hunk 2 above)
Comment (grahamlangford) [reply src/extensionConsole/pages/deployments/useAutoDeploy.ts:35]: Exactly. `isAutoDeploying` represents that the deployments are being activated and/or deployments haven't loaded yet and need to be loaded first. If i use a combination of `isLoadingDeployments` and `isAutoDeploying`, then there is a render during which the combination is `false`, which causes a UI flicker. So I keep `isAutoDeploying` set to `true` until the deployments have been loaded and either confirmed to not have any that can auto deploy or they've been deployed successfully. 

`isAttemptingAutoDeploy` prevents `activateDeployments` from running concurrently.

I'm still workshopping names. Plus I'll add some comments explaining some of this before I mark the PR ready.

I still have some edge cases I need to test thoroughly.

File: src/extensionConsole/pages/deployments/useAutoDeploy.ts (hunk 3 above)
Comment (grahamlangford) [reply src/extensionConsole/pages/deployments/useAutoDeploy.ts:32]: There are 2 versions of `deployments` at the callsite:
https://github.com/pixiebrix/pixiebrix-extension/blob/7d85d8f2cc9a78d9c4e050779d7de8bbf8de6307/src/extensionConsole/pages/deployments/DeploymentsContext.tsx#L210-L221

I'm preferring `updatedDeployments` because it's already done the work of determining which deployments should be activated. But it returns an empty array even when the definitions are still loading. Based on it's usage, I could change it to return `undefined` instead of an empty array if they're still loading. I'll make that change and eliminate the need for `isLoadingDeployments`.

I also just realized I can leverage `extensionUpdateRequired` so I'll pass that in an options arg as well.

File: src/extensionConsole/pages/deployments/useAutoDeploy.ts (hunk 4 above)
Comment (grahamlangford) [reply src/extensionConsole/pages/deployments/useAutoDeploy.ts:42]: Great catch. Thanks. And I agree. Added a note to create a ticket.

File: src/permissions/permissionsUtils.ts (hunk 5 above)
Comment (grahamlangford) [reply src/permissions/permissionsUtils.ts:77]: TIL

**Comment by grahamlangford** [comment]
@twschiller This should be ready for final review. I have not been able to successfully test it locally. See: https://pixiebrix.slack.com/archives/C02RAMRDF0R/p1706799254318139?thread_ts=1706647082.193309&cid=C02RAMRDF0R

File: src/extensionConsole/pages/mods/ModsPageLayout.tsx [hunk 7]
```diff
@@ -162,7 +164,10 @@ const ModsPageLayout: React.FunctionComponent<{
         <ModsPageToolbar tableInstance={tableInstance} />
         {/* This wrapper prevents AutoSizer overflow in a flex box container */}
         <div style={{ flex: "1 1 auto" }}>
-          {isLoading ? (
+          {/**  Do not show the Mod cards while they are being fetched or while any deployments may be automatially activated
+           * If we show the cards while deployments are auto-activated, the user will see an unnecessary Update button
```
Comment (twschiller) [inline src/extensionConsole/pages/mods/ModsPageLayout.tsx:168]: I think this is OK, but should keep an eye on impact on loading time. The check for new deployments should be quick

Comment (grahamlangford) [reply src/extensionConsole/pages/mods/ModsPageLayout.tsx:168]: It was already blocking on loading the deployments. So it's only adding in the activation. In my testing the difference in loading times appeared negligible. And this prevents the flicker of the Update button.

File: src/extensionConsole/pages/deployments/useAutoDeploy.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import notify from "@/utils/notify";
+import { type Dispatch } from "@reduxjs/toolkit";
+import { useState } from "react";
+import { useDispatch } from "react-redux";
+import useAsyncEffect from "use-async-effect";
+
+/**
+ * `true` if the deployments are still being loaded or if the deployments are being automatically deployed.
+ */
+type UseAutoDeployReturn = boolean;
```
Comment (twschiller) [inline src/extensionConsole/pages/deployments/useAutoDeploy.ts:32]: NIT: I'd prefer to see this returned in an object. I'm generally not a fan of naked booleans

File: src/extensionConsole/pages/mods/ModsPageLayout.tsx (hunk 7 above)
Comment (twschiller) [reply src/extensionConsole/pages/mods/ModsPageLayout.tsx:168]: > And this prevents the flicker of the Update button.
👍

**Comment by twschiller** [comment]
Thanks @grahamlangford  - let's cut first beta after this is merged. This will be a good one to add manual regression test(s) for

File: src/extensionConsole/pages/deployments/useAutoDeploy.ts (hunk 8 above)
Comment (grahamlangford) [reply src/extensionConsole/pages/deployments/useAutoDeploy.ts]: Done.
